torch/fx/experimental/symbolic_shapes.py

```diff
@@ -3227,6 +3227,10 @@ def issue_guard(guard: ShapeGuard) -> None:
                 self.log.warning("Failing guard allocated at: \n%s", ''.join(guard.stack.format()))
                 raise
 
+        for var, sources in self.var_to_sources.items():
+            if not symbol_to_source[var]:
+                symbol_to_source[var].extend(sources)
+
```

Comment:
@ezyang is this the right workarond for:
```py
  File "/home/jansel/pytorch/torch/_inductor/codecache.py", line 862, in load
    FxGraphCache._save_graph(key, compiled_graph, example_inputs)
  File "/home/jansel/pytorch/torch/_inductor/codecache.py", line 802, in _save_graph
    disk_compiled_graph.guards_expr = shape_env.produce_guards_expression(symints)
                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jansel/pytorch/torch/fx/experimental/symbolic_shapes.py", line 3367, in produce_guards_expression
    guards = self.produce_guards(placeholders, [LocalSource(a) for a in arg_names], ignore_static=ignore_static)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jansel/pytorch/torch/fx/experimental/symbolic_shapes.py", line 3236, in produce_guards
    issue_guard(guard)
  File "/home/jansel/pytorch/torch/fx/experimental/symbolic_shapes.py", line 3201, in issue_guard
    guard_expr = ShapeGuardPrinter(symbol_to_source, source_ref, self.var_to_sources).doprint(expr)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jansel/conda/envs/pytorch/lib/python3.11/site-packages/sympy/printing/printer.py", line 292, in doprint
    return self._str(self._print(expr))
                     ^^^^^^^^^^^^^^^^^
  File "/home/jansel/conda/envs/pytorch/lib/python3.11/site-packages/sympy/printing/printer.py", line 331, in _print
    return printmethod(expr, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jansel/conda/envs/pytorch/lib/python3.11/site-packages/sympy/printing/str.py", line 777, in _print_Relational
    return '%s(%s, %s)' % (charmap[expr.rel_op], self._print(expr.lhs),
                                                 ^^^^^^^^^^^^^^^^^^^^^
  File "/home/jansel/conda/envs/pytorch/lib/python3.11/site-packages/sympy/printing/printer.py", line 331, in _print
    return printmethod(expr, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jansel/pytorch/torch/fx/experimental/symbolic_shapes.py", line 1192, in _print_Symbol
    assert self.symbol_to_source.get(expr), (
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
AssertionError: s1 (could be from ['input0']) not in {s0: ["L['t0']"], s1: []}.  If this assert is failing, it could be due to the issue described in https://github.com/pytorch/pytorch/pull/90665
```

Reply:
According to CI it is not :P  Let me think about it.

Reply:
When I run the test script with `TORCHDYNAMO_EXTENDED_DEBUG_CREATE_SYMBOL=s1 TORCH_LOGS=dynamic python al.py`

it reveals that the naughty symbol was allocated by the pattern matcher

```
  File "/data/users/ezyang/b/pytorch/torch/_inductor/fx_passes/joint_graph.py", line 295, in joint_graph_passes 
    count += replace_random_passes(graph)                                                                      
  File "/data/users/ezyang/b/pytorch/torch/_inductor/fx_passes/replace_random.py", line 26, in replace_random_passes
    count = patterns.apply(gm)                                                                                 
  File "/data/users/ezyang/b/pytorch/torch/_inductor/pattern_matcher.py", line 1269, in apply
    entry.apply(m, graph, node)  # type: ignore[arg-type]                                
  File "/data/users/ezyang/b/pytorch/torch/_inductor/pattern_matcher.py", line 767, in apply
    self.handler(match, *match.args, **match.kwargs)                                                           
  File "/data/users/ezyang/b/pytorch/torch/_inductor/fx_passes/replace_random.py", line 139, in replace_randint 
    match.replace_by_example(replacement, [low, high, size])                           
  File "/data/users/ezyang/b/pytorch/torch/_inductor/pattern_matcher.py", line 130, in replace_by_example
    replacement = trace_fn(                                                                                    
  File "/data/users/ezyang/b/pytorch/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)                                                                               
  File "/data/users/ezyang/b/pytorch/torch/_inductor/pattern_matcher.py", line 1362, in fwd_only
    gm = make_fx(fn, select_decomp_table(), tracing_mode=mode)(*args)      
  File "/data/users/ezyang/b/pytorch/torch/fx/experimental/proxy_tensor.py", line 1111, in wrapped
    args = pytree.tree_map(wrap_fn_map[tracing_mode], args)                          
  File "/data/users/ezyang/b/pytorch/torch/utils/_pytree.py", line 900, in tree_map
    return treespec.unflatten(map(func, *flat_args))                                                           
  File "/data/users/ezyang/b/pytorch/torch/utils/_pytree.py", line 736, in unflatten
    leaves = list(leaves)                                                                                      
  File "/data/users/ezyang/b/pytorch/torch/fx/experimental/proxy_tensor.py", line 1100, in wrap_fake
    return shape_env.create_symintnode(shape_env.create_symbol(x, source, positive=None), hint=x, source=source)      
  File "/data/users/ezyang/b/pytorch/torch/fx/experimental/recording.py", line 235, in wrapper
    return fn(*args, **kwargs)                                                                                                                                                                                                
  File "/data/users/ezyang/b/pytorch/torch/fx/experimental/symbolic_shapes.py", line 2724, in create_symbol
    self.log.info(
```

If I understand correctly, the `make_fx` call in `fwd_only` is tracing the pattern, is that right? So it seems to me that you allocated a symbol for the pattern trace, but then failed to substitute it with the true match, so it leaked into real graph. So that's the real bug you should fix.

Reply:
Shouldn't the symbol already be allocated by the time it hits the pattern matcher?  The pattern matcher is just passing it through from the old op to the new op.

How is a symbol allocated?  How can I copy the already allocated symbol from the old op to the new op?

Reply:
So what the pattern matcher is doing is tracing the replacement graph, then stitching that replacement into the original graph.

I am guessing I need to copy over the symbol allocation from graph where we traced the replacement into the original one.  Where is that stored?

Reply:
Hmm. So given some pattern parametrized over some tensor inputs and some int inputs P(t0, t1, s0, s1), you have defined a replacement R(t0, t1, s0, s1), is that right? So at the time when you perform the trace, you want to feed in s0, s1 (possibly not plain symbols, but arbitrarily complicated replacements), rather than have make_fx trigger SymInt allocation on arguments.  If you are feeding fake tensors into make_fx in 1362, it's possible forcing tracing_mode="real" will do the right thing.

Reply:
This is the wrong thing to do if the replacement pattern is FX traced only once and cached everywhere; then you would want t0, t1, s0, s1 to be metavariables, and then do some sort of explicit replacement after the fact. Then, the replacement would involve stitching the FX graph in, and then rerunning fake tensor propagation on the internal nodes to give them updated metas.

Reply:
We trace the replacement every time.  It looks like right now we are feeding in the raw sympy.Exprs.